Post 0.107 cleanup & dropping API/ABI compat

slyon opened the pull request:
## Description

### Note
Please review commit-by-commit (and ignore the huge "ABI: regenerate compat check for dropped symbols" commit in `abi-compat/*.xml` -- it's autogenerated)

### Unrelated changes
This PR contains some unrelated "cleanup" changes for post 0.107 tasks.

### Primary changes
The primary changes, of dropping legacy Netplan API and ABI (in preparation for a breaking 1.0 release) start from commit "test:generator:base: Refactor to use new API".

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

slyon commented:
> * Maybe comment out these asserts to unblock autopackage tests https://github.com/canonical/netplan/blob/main/tests/integration/routing.py#L316 until we properly address the problem.

see https://github.com/canonical/netplan/pull/401

daniloegea commented on tests/generator/base.py:
```diff
@@ -269,31 +272,33 @@ def validate_generated_yaml(self, yaml_input):
+            state = netplan.State()
+            state.import_parser_results(parser)
+
+            # TODO: Allow handling of full hierarchy overrides,
```
I think this is fine for now, but we should create a task to keep track of it as it will not be possible to write tests that rely on merging without skipping this validation.

slyon replied:
:ballot_box_with_check:  FR-6338

daniloegea commented on tests/ctests/test_netplan_misc.c:
```diff
@@ -177,6 +189,107 @@ test_netplan_netdef_get_output_filename_invalid_backend(__unused void** state)
+
+    /* Check file contents */
+    FILE *fd = fopen(output_yaml, "r");
+    char file_buffer[600];
```
The unittest is failing with ASAN because this buffer is not being zeroed (on the previous test as well but it's not failing...). There is garbage in the end of string (see below). Initializing it with zeros (` = {0}`) fixes the test. The memory leak happens because the `assert_string_equal` assertion prevents `np_state` to be cleared.

```
[  ERROR   ] --- "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"" != "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"IH"
[   LINE   ] --- ../tests/ctests/test_netplan_misc.c:282: error: Failure!
```

daniloegea replied:
The RPM build is failing for the same reason (see the end of the buffer below). But there both tests are failing without ASAN, which is interesting.

```
[  ERROR   ] --- "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"" != "network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      networkmanager:
        uuid: "990548be-01ed-42d7-9f9f-cd4966b25c08"oÏwBV"
[   LINE   ] --- ../tests/ctests/test_netplan_misc.c:282: error: Failure!
```

daniloegea commented on src/util-internal.h:
```diff
@@ -34,7 +34,7 @@ wifi_frequency_24;
 wifi_frequency_5;
 
-NETPLAN_ABI void
+NETPLAN_ABI void // XXX: still used in generate.c
```
Do we need to keep the `NETPLAN_ABI` macro? We could replace it with `NETPLAN_INTERNAL`...

slyon replied:
Very true!

daniloegea reviewed (changes requested):
It's looking good so far. The only thing we must fix are the tests that are failing. They just need to be zeroed because the `\0` is missing when the strings are compared (see my comment in the test).

slyon commented:
Thanks for your review @daniloegea ! I've addressed all of your comments and got the CI to be green (except for the known tunnel/GRE failures on Jammy autopkgtest, which are unrelated).

PTAL.

daniloegea commented on .github/workflows/autopkgtest.yml:
```diff
@@ -45,7 +45,12 @@ jobs:
           rm -r debian/patches/  # clear any distro patches
-          sed -i 's|rm debian/tmp/lib/netplan/generate|# DELETED|' debian/rules
+          # usrmerge-fix
+          mkdir debian/extra
```
I believe this will start to fail when your changes get synced with Ubuntu but we can fix the CI again later.

slyon replied:
Yes. And I just uploaded 0.107.1-2 into Debian unstable. I'll use my next PR to adopt CI accordingly, once that landed.

daniloegea reviewed (approved):
lgtm!
